src/main/java/alfred/storage/AlfredStorage.java

```diff
@@ -0,0 +1,227 @@
+        this.checkValidListIndex(taskId);
+        this.taskList.remove(taskId);
+        this.saveToFile();
+    }
+
+    /**
+     * Updates the internal data state to add a Task object
+     * to the list and calls the UI to print a response.
+     *
+     * @param task Task object.
+     * @param ui   AlfredInterface object used by Alfred to handle inputs
+     *             and outputs with the user.
+     */
+    public void addTaskAndPrint(Task task, AlfredUserInterface ui) {
+        this.taskList.add(task);
+        String out = "Yes sir, I've added this task.\n";
+        out += task.toString() + "\n";
+        out += this.summarizeList();
+        ui.sandwichAndPrint(out);
+        this.saveToFile();
+    }
+
+    /**
+     * Updates the internal data state to add a Task object
+     * to the list and returns the response.
+     *
+     * @param task Task object.
+     * @param ui   AlfredInterface object used by Alfred to handle inputs
+     *             and outputs with the user.
+     */
+    public void addTask(Task task) {
+        this.taskList.add(task);
+        String out = "Yes sir, I've added this task.\n";
+        out += task.toString() + "\n";
+        out += this.summarizeList();
+        this.saveToFile();
+    }
+
+    /**
+     * Gets the string representation of a given task.
+     *
+     * @param taskId 0-indexed id of a task.
+     * @return String representation of the task.
+     * @throws InvalidIndexException If the index is out of bounds.
+     */
+    public String taskToString(int taskId) throws InvalidIndexException {
+        this.checkValidListIndex(taskId);
+        return this.taskList.get(taskId).toString();
+    }
+
+    /**
+     * Returns a formatted string that lists all matching tasks
+     * with a name that matches input text.
+     *
+     * @param text Input text to be matched with regex.
+     * @return Formatted output list of tasks.
+     */
+    public String find(String text) {
+
+        ArrayList<Task> matches = new ArrayList<Task>();
+        // iterate through arraylist to find match
+        for (int i = 0; i < this.taskList.size(); i++) {
+            Task task = this.taskList.get(i);
+            if (task.match(text)) {
+                matches.add(task);
+            }
+        }
+        if (matches.size() == 0) {
+            return "None found.";
+        } else {
+            return this.listToString(matches);
+        }
+    }
+
+    private String listToString(ArrayList<Task> taskList) {
+        String out = "";
+        for (int i = 1; i < taskList.size() + 1; i++) {
+            out += i + ". " + taskList.get(i - 1).toString() + "\n";
+        }
+        return out;
+    }
+
+    private void checkValidListIndex(int taskId) throws InvalidIndexException {
+        if (taskId >= this.taskList.size() || taskId < 0) {
+            throw new InvalidIndexException();
+        }
+    }
+
+    public String summarizeList() {
```

Comment:
Perhaps adding header comments would be easier to understand what the method does at a glance!